global/scripts/parallel.shlib

```diff
@@ -172,3 +185,74 @@ function par_addjob_stdin()
     par_jobinputarray+=("$jobinput")
 }
 
+#Add job to the current stage (FSLSUB, BUILTIN or NONE mode)
+#Inputs:
+#1. parallel mode
+#2. prallel job parameter (queue name for FSLSUB, must be 'NONE' for others) 
+#3. command with arguments
+function par_add_job_to_stage()
+{
+    local PARALLEL_MODE=$1
+    shift
+
+    #FSLSUB queue name for FSLSUB mode, should be 'NONE' for other modes.
+    local fslsub_options param=$1
+    shift
+
+	local job submit_cmd cmd=("$@")
+
+    if [ "$PARALLEL_MODE" == FSLSUB ]; then #current job is submitted, and added to fsl_job_list
+        fslsub_options=""
+        if [ "$param" != NONE ]; then 
+            fslsub_options="-q $param" 
+        fi
+        if [ -n "$par_fslsub_job_waitlist" ]; then fslsub_options="$fslsub_options -j $par_fslsub_job_waitlist"; fi
+        submit_cmd="$FSLDIR/bin/fsl_sub $fslsub_options ${cmd[@]}"
+        echo "parallel launching: $submit_cmd"
+        job=$($submit_cmd); par_fslsub_job_list+=("$job")
+    elif [ "$PARALLEL_MODE" == BUILTIN ]; then 								#current job is added to the list of local parallel jobs
+		echo "adding parallel job: $cmd"
+		par_addjob $cmd        
+    else #NONE          													#current job is executed.
+		echo "launching, parallel mode=NONE, will wait to finish: $cmd"
+		${cmd} 1> ./"$par_global_scriptname.$$.$par_global_launch_count.o.log" 2> ./"$par_global_scriptname.$$.$par_global_launch_count.e.log"
+		par_global_launch_count=$((par_global_launch_count + 1))
+		if (( $? )); then log_Err_Abort "Job failed, exiting"; fi
+    fi
+}
+
+#Finalize the list of jobs at the current stage.
+#FSLSUB: update prerequisite job waitlist for the next stage; empty current stage job list
+#BUILTIN: wait until all jobs at the current stage complete, internally, also empty current job list
+#NONE: do nothing, execution was done in par_add_job_to_stage_multimode()
+
+#Inputs:
+#1. Mode name
+#2. max jobs (BUILTIN mode only)
+function par_finalize_stage()
+{
+	#global fslsub_job_list fslsub_job_waitlist
+    local PARALLEL_MODE=$1
+    shift    
+    local temp max_jobs=$1
+    if [ -z "$max_jobs" ]; then max_jobs=-1; fi
+		
+	if [ "$PARALLEL_MODE" == "FSLSUB" ]; then 		#update the fslsub_job_waitlist with the current job list.        
+		echo "updating fsl_sub job waitlist.."
```

Comment:
Why these echoes?

Reply:
OK

Reply:
You still have another `echo` at the end of this mode.  I guess the `echo`s in BUILTIN mode are okay, as there will actually be a delay between them while things run.